subprojects/gradle/module-dependencies-graph/src/main/kotlin/com/avito/module/internal/dependencies/DependenciesGraphBuilder.kt

```diff
@@ -68,14 +49,21 @@ internal class DependenciesGraphBuilder(
                     coordinate to conf.dependencies
                         .withType(DefaultProjectDependency::class.java)
                         .filter { projectDependency ->
-                            val result = projectDependency.dependencyProject != project
-                            if (!result) {
+                            val isOtherProject = projectDependency.dependencyProject != project
+                            if (!isOtherProject) {
                                 logger.debug("Project $project depends on self in configuration ${conf.name}")
                             }
-                            result
+                            isOtherProject
+                        }
+                        .filter { projectDependency ->
+                            // Prevents cycles to itself through other configurations (e.g. test dependencies)
+                            !traversingProjects.contains(projectDependency.dependencyProject)
                         }
```

Comment:
The problem was here.
Case:
![Screenshot from 2021-03-18 20-00-57](https://user-images.githubusercontent.com/1104540/111666338-ad24c380-8824-11eb-8f83-08136aaca858.png)